Change taglist_get_tagstring to support 'unlimited' tag list size

dje29 opened the pull request:
Previous taglist_get_tagstring signature/implementation did not allow
handling of cases where inputted buffer could not contain all tags.
New implementation allocates buffer based on pre-computed size allowing to
insert all tags in the returned string.

Added get_taglist_string in qthelper to handle conversion to QString
Added TestTagList with tests for taglist_get_tagstring

Signed-off-by: Jeremie Guichard <djebrest@gmail.com>

<!-- Lines like this one are comments and will not be shown in the final output. -->
<!-- Make sure that you have read the "Contributing" section of the README and also the notes in CodingStyle. -->
<!-- If you are a collaborator, please add labels and assign other collaborators for a review. -->

### Describe the pull request:
<!-- Replace [ ] with [x] to select options. -->
- [ ] Bug fix
- [ ] Functional change
- [ ] New feature
- [x] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:
<!-- Describe your pull request in detail. -->
This is a follow up on our discussion regarding this issue in #1184 when user would use so many tags that generated string would not fit in the 1024 buffer. I have (more or less) followed @neolit123 implementation proposal. 

### Changes made:
<!-- Enumerate the changes with 1), 2), 3) etc. -->
I've described all changes in the commit message.

<!-- Ensure the test cases are updated if needed. -->
Since I do not like to write code without tests, I've added some too.

### Related issues:
<!-- Reference issues with #<issue-num>. -->
<!-- Write "Fixes #<issue-num" to notify Github that this PR fixes an issue. -->

### Additional information:
<!-- Include sample dive log or other relevant information to allow testing the change where feasible. -->

### Release note:
<!-- Describe if this change needs a release note present in CHANGELOG.md. -->
<!-- Also, please make sure to add the release note on top of the file CHANGELOG.md. -->
Note sure if this change needs a mention in the CHANGELOG.md plus I would not know what topic to head it with, is there a place where a list of topics to be used here is maintained?

### Documentation change:
<!-- If this PR makes changes to user functionality, then the documentation has to be updated too. -->
<!-- Please, briefly outline here what has changed in terms of the user experience (UX). -->
<!-- If UX changes have been made, a maintainer should apply the 'needs-documentation-change' label. -->

### Mentions:
<!-- Mention users that you want to review your pull request with @<user-name>. Leave empty if not sure. -->
@neolit123

neolit123 commented on core/dive.c:
```diff
@@ -3020,30 +3020,38 @@ void taglist_cleanup(struct tag_entry **tag_list)
 }
 
-int taglist_get_tagstring(struct tag_entry *tag_list, char *buffer, int len)
+char* taglist_get_tagstring(struct tag_entry *tag_list)
```
`char *tag...` rather than `char* tag...`

bstoeger commented on core/dive.c:
```diff
@@ -3020,30 +3020,38 @@ void taglist_cleanup(struct tag_entry **tag_list)
 	}
-	return i;
+	/* Allocate buffer and generate string with ", " separator(s) */
+	if (buffer_size > 0) {
```
Please don't make it complicated for the caller (having to handle null). If there are no tags, return an empty string. Also makes the code here more straight-forward.

neolit123 replied:
i think it's OK to return `NULL` as this is also covered in the tests.

bstoeger replied:
I think it's a really bad interface. The next person does a very reasonable
```
char *s = taglist_get_tagstring(...);
snprintf(buf, sizeof(buf), "Tags: %s\n", s);
free(s);
```
And boom  - strange and undefined behavior.

This code makes it more complicated for the caller *and* the callee. What's the advantage?

neolit123 replied:
i guess we can do that. please return `""` instead of `NULL`.
but this will also affect the test.

bstoeger replied:
My idea was to remove both `if`-conditions - if there are not tags it would simply fall back to an empty string, no?

neolit123 replied:
let's see the next iteration that @dje29 comes up with.

neolit123 commented on core/dive.c:
```diff
@@ -3020,30 +3020,38 @@ void taglist_cleanup(struct tag_entry **tag_list)
-	return i;
+	/* Allocate buffer and generate string with ", " separator(s) */
+	if (buffer_size > 0) {
+		buffer = malloc(buffer_size + 1);
```
you can just return `buffer` (`NULL`) here:
```
if (!buffer_size)
	return buffer;
```

neolit123 commented on core/dive.c:
```diff
@@ -3020,30 +3020,38 @@ void taglist_cleanup(struct tag_entry **tag_list)
+	if (buffer_size > 0) {
+		buffer = malloc(buffer_size + 1);
+	}
+	if (buffer) {
```
```
if (!buffer) {
	fprintf(stderr, "taglist_get_tagstring(): cannot allocate buffer of size: u%\n);
	return NULL;
}
```
this would extract the contents of the branch bellow.

neolit123 commented on core/dive.c:
```diff
@@ -3020,30 +3020,38 @@ void taglist_cleanup(struct tag_entry **tag_list)
+	if (buffer) {
+		last_pos = buffer;
+		tmp = tag_list;
+		while(tmp != NULL) {
```
space after `while`

neolit123 commented on core/dive.h:
```diff
@@ -264,10 +264,12 @@ struct tag_entry *taglist_added(struct tag_entry *original_list, struct tag_entr
+ * NOTE! The returned buffer must be freed once used.
  */
-int taglist_get_tagstring(struct tag_entry *tag_list, char *buffer, int len);
+char* taglist_get_tagstring(struct tag_entry *tag_list);
```
`char *`

neolit123 commented on tests/testtaglist.h:
```diff
@@ -0,0 +1,18 @@
+
+#include <QtTest>
+
+class TestTagList : public QObject{
```
add space before `{`

neolit123 commented on tests/testtaglist.cpp:
```diff
@@ -0,0 +1,55 @@
+				   ", A new tag 2"
+				   ", A new tag 3"
+				   ", A new tag 4"
+				   ", A new tag 5"));
```
there is no let's change this to:
```
QCOMPARE(QString::fromUtf8(tagstring),
         QString::fromUtf8("A new tag, "
                           "A new tag 1, "
                           "A new tag 2, "
                           "A new tag 3, "
                           "A new tag 4, "
                           "A new tag 5")
);
```
notice:
- `, ` are at the end of the strings
- only spaces for the intent as this makes the indent consistent across tab sizes

neolit123 commented:
> Note sure if this change needs a mention in the CHANGELOG.md plus I would not know what topic to head it with, is there a place where a list of topics to be used here is maintained?

there is no need for adding change log entries for internal changes that do not affect the user (e.g. code cleanup).

we don't really have a list of topics, but for this one something like `Dive-tags:` is a good one.

neolit123 commented on core/dive.c:
```diff
@@ -3020,30 +3020,38 @@ void taglist_cleanup(struct tag_entry **tag_list)
-	memset(buffer, 0, len);
+	char *buffer = NULL;
+	char *last_pos = NULL;
+	size_t buffer_size = 0;
```
let's use `unsigned int` here instead of `size_t`, this will save a cast when doing the `fprintf` later on.

neolit123 commented on tests/testtaglist.cpp:
```diff
@@ -0,0 +1,55 @@
+// SPDX-License-Identifier: GPL-2.0
+#include "testtaglist.h"
```
i would have moved the test into a separate commit, but i guess this is fine too. see if you'd prefer that?
the first commit message would need an amend in that case.

bstoeger commented:
Wouldn't `membuffer` with `put_format()` lead to much simpler code with a negligible performance impact?

neolit123 commented:
it is possible to do it with `membuffer` but i dislike the fact that `realloc` will be called here for each tag entry.

bstoeger commented:
> it is possible to do it with membuffer but i dislike the fact that realloc will be called here for each tag entry.

It won't:
```
                /* round it up to not reallocate all the time.. */
                needed = needed * 9 / 8 + 1024;
                n = realloc(b->buffer, needed);
```
In the normal case (< 1024 bytes) you'll get only a single allocation.

neolit123 commented on core/dive.c:
```diff
@@ -3020,30 +3020,38 @@ void taglist_cleanup(struct tag_entry **tag_list)
-		} else {
-			return i;
+		if (buffer_size > 0) {
+			buffer_size += 2;
```
this needs a small change. notice if a tag length is returned as `0` the size will still increment by `2`.
to avoid this store the old size and compare it instead of comparing to `0`.

neolit123 commented on core/dive.c:
```diff
@@ -3020,30 +3020,38 @@ void taglist_cleanup(struct tag_entry **tag_list)
+		while(tmp != NULL) {
+			if (last_pos != buffer) {
+				strcpy(last_pos, ", ");
+				last_pos += 2;
```
skip adding `, ` for empty tags?

neolit123 commented:
> It won't:
> In the normal case (< 1024 bytes) you'll get only a single allocation.

 i guess that's true. i leave it to @dje29 to decide if he wants to use `membuffer` with `put_format`.
examples in `core/save-xml.c`.

dirkhh commented:
I wonder why the test failed - that needs to be investigated.
Great reviews, thanks @neolit123 and @bstoeger 
I agree that I'd love to see this use a membuffer. That would result in easier, cleaner code, IMHO.
This is always the balance of being welcoming and supportive to new contributors, and pushing for consistency and adherence to our standards... @dje29 would you be willing to change the implementation to use that instead?

dje29 commented:
Here is the change with new membuffer based implementation, I did not know this utility, it is quite handy and indeed made the code much simpler. Happy to have learned about it.

- As proposed by @bstoeger now function returns null terminated empty string when no tags are passed, tests is updated accordingly
- Fixed the issue raised by @neolit123 regarding printing ", " when list contains empty tag (added test cases for that)
- I have fixed coding style mistakes and hopefully did not introduce new ones... 

Working on this contribution made me think that CONTRIBUTING.md could be extended adding
1.  Section with some coding guidelines (such has use membuffer instead of malloc, strcpy, ... when working with strings), but certainly there are others that currently resid only in your heads and can only be discovered "the hard way" :) It would simply need the habit to patch the section every time a review like the one we just did is made. I'll prepare a pull request with a proposal for such entry
2. Mention of the CHANGELOG.md use (with list of topix that can be used) since currently new contributors discover this only once they create a pull request.

bstoeger commented on core/dive.c:
```diff
@@ -3020,30 +3021,30 @@ void taglist_cleanup(struct tag_entry **tag_list)
+	 *  - empty tag list
+	 *  - tag list with empty tag only
+	 */
+	if (b.len == 0) {
```
You can replace this `if`-block by
```
mb_cstring(&b);
```
I think the membuffer interface was  intended that way anyway.

bstoeger commented on core/dive.c:
```diff
@@ -3020,30 +3021,30 @@ void taglist_cleanup(struct tag_entry **tag_list)
-			if (i > 0) {
-				strcpy(buffer + i, ", ");
-				strcpy(buffer + i + 2, tmp->tag->name);
+		if (*tmp->tag->name != '\0') {
```
You could write this as
```
if (!empty_string(tmp->tag->name))
```
This handles `name` being the null pointer gracefully.

dirkhh commented:
@dje29 thanks for the rewrite.
I'd love a PR against CONTRIBUTING.md - I'm always happy to make things easier for new contributors joining our team.
As I mentioned on the mailing list, I'm planning to cut a release today, so we have a day or two to maybe consider the review by @bstoeger  :-)

dje29 commented:
@dirkhh I preferred to fix the code as the topic was still fresh in my head, but it is indeed a good idea to stall the merge to after the current release will be done.

dirkhh commented:
Very nice - I like it. And the code is so much easier to read.
Thanks @dje29
